pkg/maintainer/proof.go

```diff
@@ -0,0 +1,130 @@
+package maintainer
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+)
+
+// AssembleTransactionProof assembles a proof that a given transaction was
+// included in the blockchain and has accumulated the required number of
+// confirmations.
+func AssembleTransactionProof(
+	transactionHash bitcoin.Hash,
+	requiredConfirmations uint,
+	bitcoinClient bitcoin.Chain,
+) (*bitcoin.Transaction, *bitcoin.Proof, error) {
+	transaction, err := bitcoinClient.GetTransaction(transactionHash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	confirmations, err := bitcoinClient.GetTransactionConfirmations(
+		transactionHash,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if confirmations < requiredConfirmations {
+		return nil, nil, fmt.Errorf(
+			"transaction confirmations number[%v] is not enough, required [%v]",
+			confirmations,
+			requiredConfirmations,
+		)
+	}
+
+	latestBlockHeight, err := bitcoinClient.GetLatestBlockHeight()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	txBlockHeight := latestBlockHeight - confirmations + 1
+
+	headersChain, err := getHeadersChain(
+		bitcoinClient,
+		txBlockHeight,
+		requiredConfirmations,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	merkleBranch, err := bitcoinClient.GetTransactionMerkleProof(
+		transactionHash,
+		txBlockHeight,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	merkleProof, err := CreateMerkleProof(merkleBranch)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create Merkle proof [%w]", err)
+	}
+
+	proof := &bitcoin.Proof{
+		MerkleProof:    merkleProof,
+		TxIndexInBlock: merkleBranch.Position,
+		BitcoinHeaders: headersChain,
+	}
+
+	return transaction, proof, nil
+}
+
+// CreateMerkleProof creates a proof of transaction inclusion in the block by
+// concatenating 32-byte-long hash values. The values are converted to the
+// little endian form. The branch of a Merkle tree leading to a transaction
+// needs to be provided. The transaction inclusion proof in hexadecimal form is
+// returned.
+func CreateMerkleProof(txMerkleBranch *bitcoin.TransactionMerkleProof) (
```

Comment:
I think there is no need to make this function public. Can we make it package-private?